feat(build): build.zig MVP: build and run functionaltests on linux

bfredl opened the pull request:
currently, this works to run a simple lua environment in the host system:

    zig build nlua0 -- hello.lua

To generate source files 

    zig build gen_sources

to build a fixed configuration for x86-64 linux only:

    zig build -Dtarget=x86_64-linux-gnu nvim_bin

# what is nlua0 ?

`nlua` is just an abbreviation for nvim flavored lua, i e lua with the extended standard library functionality added by nvim such as json, mpack and libuv. `nlua0` is the _formal_ concept of the lua code we need to run before `nvim` can be built, in order to run lua scripts which process and generate more .c code, which still need a _subset_ of these extensions. 

It is not something which will be exposed to users (they will still use `nvim -l` to run such lua code)..

# Will build.zig replace cmake?

At this point, who knows. It's worth a try I think.

# But isn't zig 0.x / pre-alpha / whatever ?

Yes, that can still be a problem, for now. We will see which project (neovim or zig) will abandon zerover first (or ever).

bfredl commented:
progress report: this can now generate `autocmd.[ch].generated.h` style prototype files.

Note: (most) dependencies are not implemented yet. so `make deps` is still needed (or having everything available as part of your system).

But from a complete `.deps` or equivalent, `zig build wip` will emit these headers as `zig-out/include/` (temporarily, in the end zig-out should just contain a fully built nvim).

clason commented on deps/vterm/include/encoding/DECdrawing.inc:
```diff
@@ -0,0 +1,36 @@
+     [0x7d] = 0x00A3,
+     [0x7e] = 0x00B7,
+   }
+ };
```
I think we should just commit these files to our mirror and be done with it.

clason commented on BRANCH_TODO.md:
```diff
@@ -0,0 +1,45 @@
+
+Runtime dependencies:
+
+- [ ] treesitter parsers
```
Those shouldn't be an issue, they compile fine with zig (in fact, that is _the_ preferred method on Windows...) For the record, here's the build flags from nvim-treesitter:
```
'zig',
      'cc',
      '-o',
      'parser.so',
      repo.files, // `src/parser.c`, `src/scanner.c` if it exists
      '-lc',
      '-Isrc',
      '-shared',
      '-Os',
```
on non-Win, add `-Wall -Wextra -fPIC -Werror=implicit-function-declaration`

bfredl commented:
This now works in a base x86-64 linux system without `base-devel` or any system /usr/include headers, etc
```
curl "https://ziglang.org/download/0.12.0/zig-linux-x86_64-0.12.0.tar.xz" > zig.tar.gz
tar xf zig.tar.gz
curl "https://codeload.github.com/bfredl/neovim/tar.gz/f39f1b9dd5c39bf32157fc56676e834fa6a5c698" > neovim.tar.gz
tar xf neovim.tar.gz
cd neovim-f39f1b9dd5c39bf32157fc56676e834fa6a5c698/
time ../zig-linux-x86_64-0.12.0/zig build -Dtarget=x86_64-linux-gnu nvim_bin
VIM=. zig-out/bin/nvim
```

Without _any_ caching build time is c a 17 seconds on a modern multi-core system, more or less the same as cmake. with a global cache of pre-fetched .tar.gz archives ( `~/.cache/zig` ) but in a fresh checkout without zig-cache, the first build is 8 seconds.

There are some caveats of course like
- built runtime files like `vim.vim` and generated docs are yet missing
- uses a hardcoded config.h for x86_64-linux-gnu only
- too much stuff gets _rebuilt_ on source/config changes, need some upstream improvements https://github.com/ziglang/zig/issues/19817

![image](https://github.com/neovim/neovim/assets/1363104/7b35a9f6-2101-499c-a073-e949a4bc6089)

616b2f commented on build.zig.zon:
```diff
@@ -0,0 +1,51 @@
+
+    .dependencies = .{
+        .ziglua = .{
+            .url = "https://github.com/natecraddock/ziglua/archive/5a7269eaddb13b118818690cea613ab25803683d.tar.gz",
```
I tried this PR just of curiosity, and had issues to reproduce, which is mainly because archive URLs are used and they are not guaranteed to have the same hash everytime (see https://github.com/orgs/community/discussions/46034)

If we could switch to git commit hashes instead e.g:
```
zig fetch --save git+https://github.com/natecraddock/ziglua#5a7269eaddb13b118818690cea613ab25803683d
```

this should solve the issue.

bfredl replied:
Great suggestion. I presume this also applies to tag based archives ?

p00f replied:
yes - the hashes are only stable if the author uploads their own tarballs on github releases

616b2f replied:
> Great suggestion. I presume this also applies to tag based archives ?

If I understood it correctly yes, because they rely on `git archive` command, and create it on the fly each time you request it and only cache it for a while to reduce the workload. And `git archive` does not guarantee that reproducibility.

bfredl commented:
I've been working on this again lately, with a focus on cross-compiling. As a first basic step I've gotten cross-compiling for s390x-linux on a x86-linux host. (also a aarch64 mac os binary _as a treat_ but as I don't have a fruit os computer at home, so I've just did a basic `nvim --help` check for now).

As a first "deliverable" of a real life improvement (it ain't much but it is honest work..) this might be used to implement a faster s390x CI job. This is implemented by emulating 390x on a x86 host including doing all of the work of compiling deps and neovim inside the qemu emulation.  Instead running the compilation directly on "bare" x86 might speed up that part of the ci job significantly (and then run the functional-tests inside qemu) .

bfredl committed:
feat(build): build.zig MVP: build and run functionaltests on linux

NEW BUILD SYSTEM!

This is a MVP implementation which supports building the "nvim" binary,
including cross-compilation for some targets.
As an example, you can build a aarch64-macos binary from
an x86-64-linux-gnu host, or vice versa

Add CI target for build.zig currently for functionaltests on linux
x86_64 only

Follow up items:

-  praxis for version and dependency bumping
-  windows 💀
-  full integration of libintl and gettext (or a desicion not to)
-  update help and API metadata files
-  installation into a $PREFIX
-  more tests and linters

bfredl commented:
working on cleaning this up. Adds a CI job doing very basic verification that building and then `./zig-out/bin/nvim --version` works without erroring out. 

I think we merge this soon after tagging 0.11 . This is not really useful to end users yet, but it it makes sense to get it into CI and then improve on it (the rest of the damn build system, including running all tests) in follow-up PR:s during the 0.12 cycle.

bfredl commented:
Did a first stab on `zig build functionaltest` . There are 26 failures and 16 errors remaining and a few more skipped tests than baseline (some missing strokes of the owl, like helptags, remaining tree-sitter parser issues, and external scripts like python).

bfredl commented on test/functional/plugin/lsp_spec.lua:
```diff
@@ -493,14 +493,17 @@ describe('LSP', function()
                 keymap = vim.fn.maparg('K', 'n', false, false)
               end)
-              return keymap:match('<Lua %d+: .+/runtime/lua/vim/lsp%.lua:%d+>') ~= nil
+              return keymap:match('<Lua %d+: .*runtime/lua/vim/lsp%.lua:%d+>') ~= nil
```
yes. The full path can be `runtime/lua/vim/lsp.lua` relative to CWD.

zeertzjq commented on runtime/doc/news.txt:
```diff
@@ -109,6 +109,12 @@ API
 
+BUILD
+
+• A Zig-based build system has been added as an alternative to Cmake. It is
+  currently limited in functionality, and Cmake remains the recommended option
```
```suggestion
• A Zig-based build system has been added as an alternative to CMake. It is
  currently limited in functionality, and CMake remains the recommended option
```

justinmk commented on runtime/gen_runtime.zig:
should this live in `src/gen` ?

bfredl replied:
The idea was that this file is a direct replacement for `runtime/CMakeLists.txt`  although this is not a requirement. `src/gen` might be a good place to centralize build-time zig code if we prefer that organization.

justinmk replied:
I see. Would a name like `build_runtime.zig` make sense? up to you

justinmk commented on src/nlua0.zig:
```diff
@@ -0,0 +1,113 @@
+
+fn init() !*Lua {
+    // Initialize the Lua vm
+    var lua = try Lua.init(std.heap.c_allocator);
```
does zig require/prefer 4-space indent or should we use 2-space indent for it?

chrboesch replied:
Zig prefers 4-space indentation and automatically inserts it when you call `zig fmt source.zig`.

justinmk commented on src/nlua0.zig:
```diff
@@ -0,0 +1,120 @@
+//! extended standard library functionality added by nvim such as json, mpack
+//! and libuv and a range of vim.* utility functions.
+//!
+//! nlua0 is an interpreter the lua code we need to run before nvim can be
```
```suggestion
//! nlua0 is an interpreter for the "bootstrap" lua code we need to run before nvim can be
```

e-kwsm commented on cmake.config/versiondef.h.in:
```diff
@@ -12,6 +12,6 @@
 
 #define NVIM_VERSION_CFLAGS "${VERSION_STRING}"
-#define NVIM_VERSION_BUILD_TYPE "$<CONFIG>"
+#define NVIM_VERSION_BUILD_TYPE "${CONFIG}"
```
With this change, `nvim --version` no longer shows `Build Type`.

alexrp commented on build.zig:
```diff
@@ -0,0 +1,423 @@
+        .HAVE_SYS_UIO_H = modernUnix,
+        .HAVE_READV = modernUnix,
+        .HAVE_DIRFD_AND_FLOCK = modernUnix,
+        .HAVE_FORKPTY = modernUnix and !tag.isDarwin(), // also on Darwin but we lack the headers :(
+        .HAVE_BE64TOH = modernUnix and !tag.isDarwin(),
```
If these are merely missing headers, but the function implementations are in libSystem, then please file a Zig issue so we can include the headers in the distribution.

bfredl commented on build.zig:
```diff
@@ -0,0 +1,423 @@
+    const api_prerelease = true;
+};
+
+// TODO(bfredl): this is for an upstream issue
```
https://github.com/ziglang/zig/pull/20051
